src/parser/mod.rs

```diff
@@ -45,3 +54,33 @@ impl<R: Read> Context<R> {
         &mut self.reader
     }
 }
+
+pub fn verify_starting_tag<R: Read>(
+    context: &mut Context<R>,
+    local_name: &'static str,
+) -> Result<(Vec<OwnedAttribute>)> {
+    //makes sure the specified starting tag is the next tag on the stream
+    //we ignore and skip all xmlevents except StartElement, Characters and EndElement
+    loop {
+        let next = context.reader.next();
+        match next {
+            Some(Ok(XmlEvent::StartElement {
+                name, attributes, ..
+            })) => {
+                ensure!(
+                    name.local_name == local_name,
+                    ErrorKind::InvalidChildElement(name.local_name, local_name)
+                );
+                return Ok(attributes);
+            }
+            Some(Ok(XmlEvent::EndElement { name, .. })) => {
+                bail!(ErrorKind::InvalidChildElement(name.local_name, local_name));
+            }
+            Some(Ok(XmlEvent::Characters(chars))) => {
+                bail!(ErrorKind::InvalidChildElement(chars, local_name));
+            }
+            Some(_) => {} //ignore other elements
```

Comment:
What kind of event would we be skipping here, shouldn't it always be a `StartElement` at this point?

Reply:
If you look at https://netvl.github.io/xml-rs/xml/reader/enum.XmlEvent.html you can see there are a lot of events possible. My "set parser config when spanwing the XmlReader" commit that I added below will at least turn XmlEvent::Whitespace and XmlEvent::CData to become XmlEvent::Character. XmlEvent::Comment is disabled by default.

This leaves us with these possible events for the Some(_) line: StartDocument, EndDocument and ProcessingInstruction. Especially StartDocument will always occur at the beginning of any xml document (even if there is no explicit StartDocument tag), so it is always emitted in our test functions. EndDocument we could handle and bail! on it, but we can also just ignore it and get "None" in the next loop iteration. Finally ProcessingInstruction sounds weird, I am not sure how they work or if they exist for gpx, so for now I thought it safe to ignore.

Reply:
Interesting, thanks for clearing that up!